### PR TITLE
test: add the loadtime tool

### PR DESCRIPTION
**Pull request by williambanfield**
## What does this pull request do?
This pull request adds the `loadtime` tool. This tool leverages the `tm-load-test` framework. Using the framework means that the only real logic that needs to be written is the logic for Tx generation. The framework does the rest. 

The tool writes a set of metadata into the transaction, including the current transaction rate, number of connections, specified size of the transaction, and the current time.

### Notes to reviewers

The payload is specified as proto and the generated definition is checked in. This has been done so that the tooling created for https://github.com/tendermint/tendermint/issues/9331 may leverage the same definition and not need to share or perform any custom logic when reading the metadata. proto also generates a reasonably compact wire representation, much more so than `json`, allowing more metadata to be able to fit within the `1kb` target specified.

### Questions to reviewers

It's a bit difficult to judge if this tool currently meets the requirements of #9330, since the documentation on the `tm-load-test` tool make it a bit confusing what the `--rate` flag represents. The documentation is as follows:

```
  -r, --rate int                        The number of transactions to generate each second on each connection, to each endpoint (default 1000)
```
However, it's a bit difficult to understand how this interacts with `send-period`
```
  -p, --send-period int                 The period (in seconds) at which to send batches of transactions (default 1)
```

does anyone know offhand?

Finally, the original issue lists a question regarding whether or not the data should be contained with the 'key' or the 'value' of the `kvstore` transaction. Currently, this code does not prepend a `<key>=`, which results in the transaction being used as both they key and the value in the store. Is the primary concern performance of the kv store? 

closes: https://github.com/tendermint/tendermint/issues/9330

#### PR checklist

- [ ] Tests written/updated, or no tests needed
- [ ] `CHANGELOG_PENDING.md` updated, or no changelog entry needed
- [ ] Updated relevant documentation (`docs/`) and code comments, or no
      documentation updates needed

### CODE REVIEW

**Comment by sergio-mena** [comment]
> Is the primary concern performance of the kv store?
In our conversation last week, the concern was that if we don't limit the keys used somehow, we'll end up with a gigantic state.

File: test/loadtime/Makefile [hunk 1]
```diff
@@ -0,0 +1,32 @@
...
+.PHONY: build
+
+check-proto-gen-deps:
+ifeq (,$(shell which protoc))
+	$(error "protoc is required for Protobuf generation. See instructions for your platform on how to install it.")
+endif
+ifeq (,$(shell which protoc-gen-go))
+	$(error "protoc-gen-go is required for Protobuf generation. See instructions for your platform on how to install it.")
+endif
+.PHONY: check-proto-format-deps
```
Comment (sergio-mena) [inline test/loadtime/Makefile:15]:
```suggestion
.PHONY: check-proto-gen-deps
```

File: test/loadtime/main.go [hunk 2]
```diff
@@ -0,0 +1,86 @@
...
+
+	"github.com/informalsystems/tm-load-test/pkg/loadtest"
+	"github.com/tendermint/tendermint/test/loadtime/payload"
+	"google.golang.org/protobuf/proto"
+	"google.golang.org/protobuf/types/known/timestamppb"
+)
+
+// payloadSizeBytes holds the maximum encoded size of the payload message, not
+// counting the 'Padding' field.
+// This value is used to calculated how much padding should be included in the
```
Comment (sergio-mena) [inline test/loadtime/main.go:15]:
```suggestion
// This value is used to calculate how much padding should be included in the
```

File: test/loadtime/main.go [hunk 3]
```diff
@@ -0,0 +1,86 @@
...
+	loadtest.Run(&loadtest.CLIConfig{
+		AppName:              "loadtime",
+		AppShortDesc:         "Generate timestamped transaction load.",
+		AppLongDesc:          "loadtime generates transaction load for the purpose of measuring the end-to-end latency of a transaction from submission to execution in a Tendermint network.",
+		DefaultClientFactory: "loadtime-client",
+	})
+}
+
+func (f *ClientFactory) ValidateConfig(cfg loadtest.Config) error {
+	if cfg.Size < payloadSizeBytes {
```
Comment (sergio-mena) [inline test/loadtime/main.go:55]: Very minor. Just an idea.
Since the payload has constant size, rather than hard-coding `payloadSizeBytes` (which will be error prone to update), why not having an "Init" function that populates `Payload` with 1 padding byte, and marshalls it to measure the payload size?

**Review by sergio-mena — APPROVED** [review]
Cool job! I guess this means we can drop the ad-hoc modifications to the load test tool we did for testnet-testing v0.35.x

File: test/loadtime/main.go (hunk 3 above)
Comment (williambanfield) [reply test/loadtime/main.go]: Good call. Done.

**Comment by williambanfield** [comment]
I've updated the transaction generation to prepend the tx with `a=` so that all transactions overwrite the same key in the kvstore. We don't care about storing these in the kvstore at all, what we care about is storing them in the block chain. Therefore, using the same key every time is fine.

**Comment by sergio-mena** [comment]
> Therefore, using the same key every time is fine.

I agree, until a concrete use case justifies using more than one key

**Comment by thanethomson** [comment]
> it's a bit difficult to understand how this interacts with `send-period`

The send period is an integer value whose smallest possible value is 1 (i.e. 1 second) that controls how frequently we send "bursts" of transactions at the transaction rate.

For instance, if you set the send period to 1s and rate to 100, the tool will try to send 100 transactions every 1 second. If you set the send period to 2s and rate to 100, the tool will try to send 100 transactions every 2 seconds.

With this you could simulate "bursty" traffic, by, for example, setting the send period to 60s and rate to 10,000. This would attempt to send a burst of 10k txs once every 60s.

If you want to see the mechanics of how this works under the hood, check out the [`sendLoop` function](https://github.com/informalsystems/tm-load-test/blob/e26e39e2b8f64c156bc97b6c23ae4ac0e682dc83/pkg/loadtest/transactor.go#L166).

In short, for these tests, simply set the send period to 1 :slightly_smiling_face: 

> I've updated the transaction generation to prepend the tx with `a=` so that all transactions overwrite the same key in the kvstore. We don't care about storing these in the kvstore at all, what we care about is storing them in the block chain. Therefore, using the same key every time is fine.

I'd be concerned that using this approach would potentially cause problems as discovered in #3865. I'd recommend that we rather cycle through a reasonable number of keys, e.g. using hex encoding to run from `0000` through `FFFF` or something.

**Comment by williambanfield** [comment]
> I'd be concerned that using this approach would potentially cause problems as discovered in https://github.com/tendermint/tendermint/issues/3865.

I'm not sure I see how these are related. If a duplicate transaction can be resubmitted, then this can happen irrespective of if the kvstore key is the same or not. That behavior appears more to be a defect in the mempool and mempool cache / recheck interaction than in the store itself.

**Comment by thanethomson** [comment]
> I'm not sure I see how these are related. If a duplicate transaction can be resubmitted, then this can happen irrespective of if the kvstore key is the same or not. That behavior appears more to be a defect in the mempool and mempool cache / recheck interaction than in the store itself.

Yeah I suppose I was mainly concerned with how duplicate transactions would be handled, but if we can ensure that at least the values of each key/value pair are unique that'd be fine too. Then we can definitely use the same key :slightly_smiling_face:

**Comment by thanethomson** [comment]
Very cool, great work here @williambanfield! I look forward to seeing it in action. Have you tested it out against any running nodes yet?

Also, I think we should backport this to the `v0.37.x` branch, right?

**Comment by williambanfield** [comment]
@thanethomson I tested it locally against a running node and it seem to perform just fine. 

Yes, I'll backport to `v0.37` and `v0.34`
